Histogram observer

dskhudia opened the pull request:
Summary: Adding histogram observer. It's ported from caffe2/quantization/server/norm_minimization.cc

Differential Revision: D16431012

jerryzh168 commented on torch/quantization/observer.py:
```diff
@@ -2,49 +2,52 @@
+    def calculate_qparams(self, **kwargs):
+        pass
+
+    def _calculate_qparams(self, min_val, max_val):
```
why do we put this here? maybe move to MinMaxObserver?

jerryzh168 replied:
oh it's used in other class as well.

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -369,6 +368,19 @@ def test_fuse_module_eval(self):
         self.assertEqual(type(testMod.sub2.bn), torch.nn.BatchNorm2d,
                          "Non-fused submodule BN")
 
+class ObserverTest(QuantizationTestCase):
```
Can we check that the implementation here is identical to the caffe2 implementation?

dskhudia replied:
The one I have added matches caffe2. Is there a way to call caffe2 implementation from here?

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -63,13 +77,302 @@ def calculate_qparams(self):
+        norm_min = float("inf")
+
+        while alpha < beta:
+            #pdb.set_trace()
```
Please remove comment

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -63,13 +77,302 @@ def calculate_qparams(self):
+        #pdb.set_trace()
+        new_min = self.min_val + bin_width * start_bin
+        new_max = self.min_val + bin_width * (end_bin + 1)
+        print("daya {} {} {} {}".format(new_min, new_max, start_bin, end_bin))
```
Please remove print

dskhudia replied:
:P

raghuramank100 reviewed (changes requested):
First pass, will update with additional comments later.

dskhudia committed:
Histogram observer (#23191)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/23191

Adding histogram observer. It's ported from caffe2/quantization/server/norm_minimization.cc

Differential Revision: D16431012

fbshipit-source-id: 4413eebf1585be96e659aa0a22461ac5d139ee6a

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -63,13 +76,298 @@ def calculate_qparams(self):
+        self.max_val = None
+
+    @staticmethod
+    def _get_norm(delta_begin, delta_end, density, norm_type):
```
Lets simplify here. Since we support only L2 norm, rewrite as:
```python
def _get_norm(delta_begin, delta_end, density):
            return (delta_end * delta_end * delta_end
                - delta_begin * delta_begin * delta_begin
            ) / 3*density
 ```

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -63,13 +76,298 @@ def calculate_qparams(self):
+            if dst_bin_cnt < src_bin_count:
+                dst_histogram[dst_bin2] += src_bin_count - dst_bin_cnt
+
+    def forward(self, x):
```
Lets go with a simpler implementation, the one here is too complex and hard to reason out. Here is a suggestion:
``` python
def forward(self,x):
     if self.min_val is None or self.max_val is None or self.histogram is None:
            self.min_val = 3.0* torch.min(0,  torch.min(x))
            self.max_val = 3.0* torch.max(0, torch.max(x))
           # provide sufficient dynamic range for capturing statistics
          self.histogram = torch.histc(x, self.bins, self.min_val, self.max_val)
     else:
          new_histogram = torch.histc(x, self.bins,self.min_val, self.max_val)
          self.histogram = self.histogram + new_histogram
```

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -63,13 +76,298 @@ def calculate_qparams(self):
+        new_max = self.min_val + bin_width * (end_bin + 1)
+        return new_min, new_max
+
+    def _combine_histograms(
```
Can delete this function

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -63,13 +76,298 @@ def calculate_qparams(self):
+            ) / 3
+        return density * norm
+
+    def _include_zero(self):
```
Can delete this function
